awsplugins/ec2/ec2.go

```diff
@@ -9,35 +9,115 @@
+	ImageID                 string
+	InstanceID              string
+	InstanceType            string
+	KernelID                string
+	PendingTime             string
+	PrivateIP               string
+	RamdiskID               string
+	Region                  string
+	Version                 string
+}
+
+//Init activates EC2Plugin at runtime.
 func Init() {
 	if plugins.InstancePluginMetadata != nil && plugins.InstancePluginMetadata.EC2Metadata == nil {
 		addPluginMetadata(plugins.InstancePluginMetadata)
 	}
 }
 
 func addPluginMetadata(pluginmd *plugins.PluginMetadata) {
-	session, e := session.NewSession()
-	if e != nil {
-		logger.Errorf("Unable to create a new ec2 session: %v", e)
-		return
+	var instanceData Metdata
+	imdsURL := "http://169.254.169.254/latest/"
+
+	client := &http.Client{
+		Transport: http.DefaultTransport,
+	}
+
+	token, err := getToken(imdsURL, client)
+	if err != nil {
+		logger.Errorf("Unable to fetch EC2 instance metadata token fallback to IMDS V1: %v", err)
 	}
-	client := ec2metadata.New(session)
-	doc, err := client.GetInstanceIdentityDocument()
+
+	resp, err := getMetadata(imdsURL, client, token)
 	if err != nil {
 		logger.Errorf("Unable to read EC2 instance metadata: %v", err)
 		return
 	}
 
-	pluginmd.EC2Metadata = &plugins.EC2Metadata{InstanceID: doc.InstanceID, AvailabilityZone: doc.AvailabilityZone}
+	buf := new(bytes.Buffer)
+	if _, err := buf.ReadFrom(resp.Body); err != nil {
+		logger.Errorf("Error while reading data from response buffer: %v", err)
+	}
+	metadata := buf.String()
+
+	if err := json.Unmarshal([]byte(metadata), &instanceData); err != nil {
+		logger.Errorf("Error while unmarshal operation: %v", err)
+	}
+
+	pluginmd.EC2Metadata = &plugins.EC2Metadata{InstanceID: instanceData.InstanceID, AvailabilityZone: instanceData.AvailabilityZone}
 	pluginmd.Origin = Origin
 }
+
+// getToken fetches token to fetch EC2 metadata
+func getToken(imdsURL string, client *http.Client) (string, error) {
+	ttlHeader := "X-aws-ec2-metadata-token-ttl-seconds"
+	defaultTTL := "60"
+	tokenURL := imdsURL + "api/token"
+
+	req, _ := http.NewRequest("PUT", tokenURL, nil)
+	req.Header.Add(ttlHeader, defaultTTL)
+	resp, err := client.Do(req)
+	if err != nil {
+		return "fallback", err
+	}
+
+	buf := new(bytes.Buffer)
+	if _, err := buf.ReadFrom(resp.Body); err != nil {
+		logger.Errorf("Error while reading data from response buffer: %v", err)
+	}
+	token := buf.String()
+
+	return token, err
+}
+
+// getMetadata fetches instance metadata
+func getMetadata(imdsURL string, client *http.Client, token string) (*http.Response, error) {
+	var metadataHeader string
+	metadataURL := imdsURL + "dynamic/instance-identity/document"
+
+	if token != "fallback" {
+		metadataHeader = "X-aws-ec2-metadata-token"
+	}
+
+	req, _ := http.NewRequest("GET", metadataURL, nil)
+	if token != "fallback" {
+		req.Header.Add(metadataHeader, token)
+	}
+
+	resp, err := client.Do(req)
+	if err != nil {
+		return resp, err
+	}
+
+	return resp, err
```

Comment:
can you just replace this whole block with:
```go
return client.Do(req)
```

Reply:
Done.